### ISSUE TITLE
`generator_mode=false` and `keep_doctype=false` generates empty output when <html> is missing in the source page

### ISSUE THREAD

**Issue by copy** [dmbaturin/soupault#27]
To reproduce:

1. `soupault --init`
2. Edit `soupault.conf` to `keep_doctype = false` and `generator_mode = false` (the latter may not be necessary)
3. `soupault`
4. Observe that the generate file `build/index.html` is empty (expected to be `site/index.html` with the doctype from `soupault.conf` added)
5. If one adds a doctype to `site/index.html` or sets `keep_doctype = true` it works as expected

Reproduced using 2.5.0 and current master.

**Comment by dmbaturin** [comment]
It happens when _three_ circumstances align:

* `generator_mode = false`
* `keep_doctype = false`
* The page lacks the `<html>` element.

In other words, this is a negative effect of `keep_doctype = false` on incomplete pages.

I have to admit I might have never tested `generator_mode = false` with incomplete pages (since its intended use case is post-processing complete HTML pages), so this bug likely exists since the inception of the `keep_doctype` option.

While this likely doesn't affect the intended use case, it's still counterintuitive and user-hostile behavior. I'll look into it.

**Comment by dmbaturin** [comment]
```
$ grep keep_doctype soupault.conf 
  keep_doctype = false
  # With keep_doctype=false, soupault will replace existing declarations with it too
$ grep generator_mode soupault.conf 
  generator_mode = false
  # If you use "generator_mode = false", this file is not required.


$ cat site/index.html 
<html><body>
<h1>Welcome!</h1><p>Welcome to my homepage. It's under construction.</p><p>Very incomplete!</p>
</body></html>

$ ~/devel/soupault/_build/default/src/soupault.exe 
[INFO] Running in HTML processor mode, not using the page template
[INFO] Processing page site/index.html
[INFO] Processing widget generator-meta on page site/index.html
[INFO] Processing widget test on page site/index.html
[INFO] Processing widget no-blink on page site/index.html
[INFO] Processing widget page-title on page site/index.html

$ cat build/index.html 
<!DOCTYPE html>
<html>
 <head>
  <meta name="generator" content="soupault">
 </head>
 <body>
  <h1>
   Welcome!
  </h1>
  <p>
   Welcome to my homepage. It's under construction.
  </p>
  <p>
   Very incomplete!
  </p>
 </body>
</html>

```

**Comment by dmbaturin** [comment]
https://github.com/dmbaturin/soupault/blob/master/src/soupault.ml#L80-L80

Indeed, it looks like I completely forgot about the post processor mode when I wrote that code. The else clause assumes that a page is always well-formed, even though it's no guaranteed.

I'll make it just output the HTML as is, since a doctype would be a useless "doctype decoration" in that case.

**Comment by copy** [comment]
To elaborate on my scenario, I'm mixing complete HTML pages and partial ones (some are generated by tools, some are hand-written). I had hoped that soupault could fix the partial ones (add the missing doctype, title and `<meta charset>`; I don't care if html/head/body are missing), but I understand that that's a bit of an odd use case.

I think I can work around the issue on my side — either running soupault twice or pre-processing the partial files separately.

**Comment by dmbaturin** [comment]
>I'm mixing complete HTML pages and partial ones

You need to _enable_ generator mode then. The HTML processor mode is there only to allow operating without any template files.

The generator mode isn't incompatible with having complete pages—it's one of the "selling points" of soupault and one of the things that sets it apart from other SSGs! If a page has an `<HTML>` element, it's excluded from the templating process and only processed by widgets.
See https://soupault.app/reference-manual/#partial-and-complete-pages

Just make an appropriate template that has all the boilerplate you need, and partial pages will be made complete using that template, while complete pages will be processed as a whole.
https://soupault.app/reference-manual/#page-templates

**Comment by dmbaturin** [comment]
Since the issue as described is fixed by [dff9a2f](https://github.com/dmbaturin/soupault/commit/dff9a2f006e845dce1677c6ba5ccf25939bbb694), I'm closing this.

If you have questions regarding configuration or usage, please write to the mailing list.
